File: test-suite-spring-java/src/test/java/io/micronaut/openapi/spring/TestControllerTest.java
package io.micronaut.openapi.spring;

import com.fasterxml.jackson.core.JsonProcessingException;
import io.micronaut.openapi.OpenApiUtils;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.parameters.Parameter;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.web.client.RestClient;

import java.util.List;

import static io.micronaut.openapi.spring.TestConfig.APP_NAME;
import static io.micronaut.openapi.spring.TestConfig.APP_VERSION;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

@ActiveProfiles("test")
@SpringBootTest(useMainMethod = SpringBootTest.UseMainMethod.ALWAYS, classes = {
    WebConfig.class,
    TestConfig.class,
    Application.class,
}, webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
class TestControllerTest {

    @Autowired
    RestClient restClient;

    @Test
    void springOpenApiPathTest() throws JsonProcessingException {
        var result = restClient.get()
            .uri("/swagger/" + APP_NAME + '-' + APP_VERSION + ".yml")
            .retrieve()
            .body(String.class);

        var openApi = OpenApiUtils.getYamlMapper().readValue(result, OpenAPI.class);
        assertNotNull(openApi.getInfo());
        assertEquals(APP_VERSION, openApi.getInfo().getVersion());
        assertNotNull(openApi.getPaths());

        var userSchema = openApi.getComponents().getSchemas().get("User");
        assertNotNull(userSchema);

        var createPostOp = openApi.getPaths().get("/create").getPost();

        assertEquals("Create post op summary.", createPostOp.getSummary());
        assertEquals("Create post op summary. Operation post description.", createPostOp.getDescription());
        assertNotNull(createPostOp.getRequestBody());

        var mediaType = createPostOp.getRequestBody().getContent().get("application/json");
        assertNotNull(mediaType);
        assertNotNull(mediaType.getSchema());
        assertEquals("#/components/schemas/User", mediaType.getSchema().get$ref());
        assertNotNull(createPostOp.getResponses());
        assertNotNull(createPostOp.getResponses().get("200"));
        assertEquals("created post user", createPostOp.getResponses().get("200").getDescription());

        var createPatchOp = openApi.getPaths().get("/create").getPatch();

        assertEquals("Create patch op summary.", createPatchOp.getSummary());
        assertEquals("Create patch op summary. Operation patch description.", createPatchOp.getDescription());

        mediaType = createPatchOp.getRequestBody().getContent().get("application/json");
        assertNotNull(mediaType);
        assertNotNull(mediaType.getSchema());
        assertEquals("#/components/schemas/User", mediaType.getSchema().get$ref());
        assertNotNull(createPatchOp.getResponses());
        assertNotNull(createPatchOp.getResponses().get("202"));
        assertEquals("createPatch 202 response", createPatchOp.getResponses().get("202").getDescription());

        var userIdOp = openApi.getPaths().get("/{userId}").getGet();

        assertNotNull(userIdOp);

        var params = userIdOp.getParameters();
        assertNotNull(params);
        assertEquals(2, params.size());

        var userIdParam = getParamByName("userId", params);
        assertNotNull(userIdParam);
        assertEquals("userId", userIdParam.getName());
        assertEquals("path", userIdParam.getIn());
        assertTrue(userIdParam.getRequired());
        assertNotNull(userIdParam.getSchema());
        assertEquals("string", userIdParam.getSchema().getType());

        var ageParam = getParamByName("age", params);
        assertNotNull(ageParam);
        assertEquals("age", ageParam.getName());
        assertEquals("query", ageParam.getIn());
        assertNotNull(ageParam.getSchema());
        assertNull(ageParam.getRequired());
        assertEquals("integer", ageParam.getSchema().getType());
        assertEquals("int32", ageParam.getSchema().getFormat());
        assertEquals(123, ageParam.getSchema().getDefault());
        assertTrue(ageParam.getSchema().getNullable());
    }

    private Parameter getParamByName(String name, List<Parameter> params) {
        return params.stream()
            .filter(p -> name.equals(p.getName()))
            .findFirst()
            .orElse(null);
    }
}

File: test-suite-spring-kotlin-kapt/src/test/kotlin/io/micronaut/openapi/spring/TestControllerTest.kt
package io.micronaut.openapi.spring

import tools.jackson.core.JacksonException
import io.micronaut.openapi.OpenApiUtils
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.parameters.Parameter
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.web.client.RestClient

@ActiveProfiles("test")
@SpringBootTest(
    useMainMethod = SpringBootTest.UseMainMethod.ALWAYS,
    classes = [
        WebConfig::class,
        TestConfig::class,
        Application::class,
    ],
    webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT,
)
class TestControllerTest {

    @Autowired
    lateinit var restClient: RestClient

    @Disabled
    @Test
    @Throws(JacksonException::class)
    fun springOpenApiPathTest() {
        val result = restClient.get()
            .uri("/swagger/" + TestConfig.APP_NAME + '-' + TestConfig.APP_VERSION + ".yml")
            .retrieve()
            .body(String::class.java)

        val openApi = OpenApiUtils.getYamlMapper().readValue(result, OpenAPI::class.java)
        assertNotNull(openApi.info)
        assertEquals(TestConfig.APP_VERSION, openApi.info.version)
        assertNotNull(openApi.paths)

        val userSchema = openApi.components.schemas["User"]!!
        assertNotNull(userSchema)

        val createPostOp = openApi.paths["/create"]!!.post

        assertEquals("Create post op summary.", createPostOp.summary)
        assertEquals("Create post op summary. Operation post description.", createPostOp.description)
        assertNotNull(createPostOp.requestBody)

        var mediaType = createPostOp.requestBody.content["application/json"]
        assertNotNull(mediaType)
        assertNotNull(mediaType!!.schema)
        assertEquals("#/components/schemas/User", mediaType.schema.`$ref`)
        assertNotNull(createPostOp.responses)
        assertNotNull(createPostOp.responses["200"])
        assertEquals("created post user", createPostOp.responses["200"]!!.description)

        val createPatchOp = openApi.paths["/create"]!!.patch

        assertEquals("Create patch op summary.", createPatchOp.summary)
        assertEquals("Create patch op summary. Operation patch description.", createPatchOp.description)

        mediaType = createPatchOp.requestBody.content["application/json"]
        assertNotNull(mediaType)
        assertNotNull(mediaType!!.schema)
        assertEquals("#/components/schemas/User", mediaType.schema.`$ref`)
        assertNotNull(createPatchOp.responses)
        assertNotNull(createPatchOp.responses["202"])
        assertEquals("createPatch 202 response", createPatchOp.responses["202"]!!.description)

        val userIdOp = openApi.paths["/{userId}"]!!.get

        assertNotNull(userIdOp)

        val params = userIdOp.parameters
        assertNotNull(params)
        assertEquals(2, params.size)

        val userIdParam = getParamByName("userId", params)
        assertNotNull(userIdParam)
        assertEquals("userId", userIdParam!!.name)
        assertEquals("path", userIdParam.getIn())
        assertTrue(userIdParam.required)
        assertNotNull(userIdParam.schema)
        assertEquals("string", userIdParam.schema.type)

        val ageParam = getParamByName("age", params)
        assertNotNull(ageParam)
        assertEquals("age", ageParam!!.name)
        assertEquals("query", ageParam.getIn())
        assertNotNull(ageParam.schema)
        assertNull(ageParam.required)
        assertEquals("integer", ageParam.schema.type)
        assertEquals("int32", ageParam.schema.format)
        assertEquals(123, ageParam.schema.default)
    }

    private fun getParamByName(name: String, params: List<Parameter?>): Parameter? =
        params.firstOrNull { name == it!!.name }
}
